source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
+     - Contains information about the BSON data that it decodes to a Java
+       object.
+
+The ``getEncoderClass()`` method returns a class instance of the Java class
+since Java cannot infer the type due to due to type erasure.
+
+.. _codecs-powerstatus-codec:
+
+See the following ``PowerStatusCodec`` class that implements ``Codec`` in
+order to convert the Java ``String`` values "on" and "off" to
+corresponding BSON boolean values.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/PowerStatusCodec.java
+   :start-after: start class
+   :end-before: end class
+   :language: java
+   :dedent:
+
+You can add an instance of your custom Codec to the ``CodecRegistry`` which
+contains a mapping between your ``Codec`` and the Java object type to which
+it applies. Continue to the :ref:`CodecRegistry <codecs-codecregistry>`
+section of this page to see how you can include your Codec.
+
+For more information about the classes and interfaces in this section, see the
+following API documentation:
+
+- :java-docs:`Codec <apidocs/bson/org/bson/codecs/Codec.html>`
+- :java-docs:`BsonWriter <apidocs/bson/org/bson/BsonWriter.html>`
+- :java-docs:`BsonBinaryWriter <apidocs/bson/org/bson/BsonBinaryWriter.html>`
+- :java-docs:`EncoderContext <apidocs/bson/org/bson/codecs/EncoderContext.html>`
+- :java-docs:`BsonReader <apidocs/bson/org/bson/BsonReader.html>`
+- :java-docs:`DecoderContext <apidocs/bson/org/bson/codecs/DecoderContext.html>`
+- :java-docs:`BsonBinaryReader <apidocs/bson/org/bson/BsonBinaryReader.html>`
+
+.. _codecs-codecregistry:
 
 CodecRegistry
 -------------
 
-- Description
-- Example of setting Codec list
-- Example of getting a Codec
+A ``CodecRegistry`` is an immutable collection of ``Codec`` instances that
+encode and decode the Java classes they specify. You can use any of the
+following ``CodecRegistries`` class static factory methods to construct a
+``CodecRegistry`` from the ``Codec`` instances contained in the associated
+types:
+
+- ``fromCodecs()``
+- ``fromProviders()``
+- ``fromRegistries()``
+
+The following code snippet shows how to construct a ``CodecRegistry`` using
+the ``fromCodecs()`` method:
+
+.. code-block:: java
+   :copyable: true
+
+   CodecRegistry codecRegistry = CodecRegistries.fromCodecs(new IntegerCodec(), new PowerStatusCodec());
+
+In the example above, we assign the ``CodecRegistry`` the following ``Codec``
+implementations:
+
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`,
+  a ``Codec`` that converts Integers and is part of the BSON package.
+- :ref:`PowerStatusCodec <codecs-powerstatus-codec>`, our sample ``Codec``
+  that converts certain Java strings to BSON booleans.
+
+You can retrieve the Codecs from the ``CodecRegistry`` instance from the
+prior example using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   Codec<String> powerStatus = codecRegistry.get(String.class);
+   Codec<Integer> integerCodec = codecRegistry.get(Integer.class);
```

Comment:
Since this can throw, is it worth wrapping in a try/catch?

Reply:
I think it could either:
- show the try/catch and omit the description underneath because that could be made apparent through the error message
- keep it short to just demonstrate the syntax, and then explain a potential runtime exception as it is currently structured

Is there one way that's better than the other? I don't think we've done much with handling specific exceptions in our prior documentation.

Reply:
When I did the indexes page, I wrapped some of the index creation methods in try/catch since they can throw. I'm unsure what the best approach is, perhaps it is to wrap in try/catch and catch the specific exception. If there is more than one, perhaps we add a note in that case?

Reply:
I don't think we should try to show and explain all the potential exceptions, other than those that are ones that might expect a user to frequently encounter. A few reasons for this:

- Java IDEs like Eclipse and IntelliJ generate try/catch block code for any of the potential exception types.
- There are a lot of different exception types in the methods we document, so describing each of them in detail beyond what's offered in the API documentation we link to has all the disadvantages we've discussed in denormalizing API documentation in reference documentation applies.